thetis/timeintegrator.py

```diff
@@ -433,6 +433,113 @@ def advance(self, t, update_forcings=None):
+        }
+
+        # Create functions to hold the values of previous time step
+        self.fields_old = {tracer: {} for tracer in self.fields}
+        for tracer, fields in self.fields.items():
+            for k in sorted(fields):
+                if fields[k] is not None:
+                    if isinstance(fields[k], Function):
+                        self.fields_old[tracer][k] = Function(
+                            fields[k].function_space())
+                    elif isinstance(fields[k], Constant):
+                        self.fields_old[tracer][k] = Constant(fields[k])
+
+        if semi_implicit:
+            solutions_nl = self.solutions_old
+        else:
+            solutions_nl = self.solutions
+
+        # Forms
+        theta_const = Constant(theta)
+        self.forms = {
+            tracer:
+                equation.mass_term(self.solutions[tracer]) - equation.mass_term(self.solutions_old[tracer])
+                - self.dt_const*(
+                    theta_const*equation.residual('all', self.solutions[tracer], solutions_nl[tracer], self.fields[tracer], self.fields[tracer], bnd_conditions[tracer])
+                    + (1-theta_const)*equation.residual('all', self.solutions_old[tracer], self.solutions_old[tracer], self.fields_old[tracer], self.fields_old[tracer], bnd_conditions[tracer])
+                )
+            for tracer, equation in self.equations.items()
+        }
+        self.update_solver()
+
+    def update_solver(self):
+        """Create solver objects"""
+        problems = AttrDict({
+            tracer: NonlinearVariationalProblem(F, self.solutions[tracer])
+            for tracer, F in self.forms.items()
+        })
+        if self.solver_parameters.get('pc_type') == 'lu':
+            self.solver_parameters['mat_type'] = 'aij'
+        self.solvers = {
+            tracer: NonlinearVariationalSolver(problem,
+                                               solver_parameters=self.solver_parameters,
+                                               options_prefix=self.name)
+            for tracer, problem in problems.items()
+        }
+
+    def initialize(self, solutions):
+        """Assigns initial conditions to all required fields."""
+        for tracer, solution in solutions.items():
+            self.solutions_old[tracer].assign(solution)
+            for k in sorted(self.fields_old[tracer]):
+                self.fields_old[tracer][k].assign(self.fields[tracer][k])
+
+    def advance(self, t, update_forcings=None):
+        """Advances equations for one time step."""
+        if update_forcings is not None:
+            update_forcings(t + self.dt)
+        for tracer, solution in self.solutions.items():
+            self.solutions_old[tracer].assign(solution)
+
+        for it in range(self.iterations):
+            for tracer, solver in self.solvers.items():
+                with timed_stage(f"{tracer} solve"):
+                    solver.solve()
```

Comment:
Just realised that the order cannot be trusted here. I guess we could use an `OrderedDict`?

Reply:
Sure, sounds like a good solution as the order of tracers matter.